### PR TITLE
A first test for Chapel + MPI

### PR DESCRIPTION
**Pull request by npadmana**
@ben-albrecht -- here is a first Hello world. 

There are two .good files, one for running with mpirun and single-locale Chapel (-spmd.good) and the other for multi-locale Chapel (-multilocale.good). I ran both cases with 4 ranks/locales. 

The ordering should be reproducible...

We may also want to run a case with just one rank, to make sure everything works there too. I can generate the good files, once you tell me how to set up these cases.

### CODE REVIEW

**Comment by ben-albrecht** [comment]
Thanks for the contribution @npadmana

Some TODOs for the core-team to get these tests up an running:

* [ ] Add test scripts for `mpicc` && `gasnet.mpicc`
    * Both test configurations run over `test/release`
    * `mpicc` runs with `CHPL_COMM=none` over `test/modules/packages/mpi/spmd`
    * `gasnet.mpicc` runs with `CHPL_COMM=gasnet` / `CHPL_COMM_SUBSTRATE=mpi` over `test/modules/packages/mpi/multilocale`
* [ ] Set up Jenkins nightly testing to fire off test scripts
* [ ] Modify `CHPL_LAUNCHER=mpirun` such that MPI module is compatible for `-nl >1`

File: test/modules/packages/mpi/multilocale/hello-chapel-nl-1/SKIPIF [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+compiler=os.environ['CHPL_TARGET_COMPILER']
+retval = (compiler !="mpi-gnu") & (compiler!="cray-prgenv-gnu")
+
+# Only support fifo 
+retval |= (os.environ['CHPL_TASKS']!="fifo")
+
+# Only support GASNET+MPI 
+retval |= (os.environ['CHPL_COMM']!="gasnet")
+
+if ('CHPL_COMM_SUBSTRATE' not in os.environ) :
```
Comment (ben-albrecht) [inline test/modules/packages/mpi/multilocale/hello-chapel-nl-1/SKIPIF]: After #4216 is merged, this env var should always be defined during a `start_test` run.

**Comment by npadmana** [comment]
@ben-albrecht -- thinking about this a little more, here is a possible update to the mpirun launcher that includes the functionality we want.

`CHPL_MPI_NUM_RANKS=4 ./a.out ` yields
`Hello, Chapel! This is MPI rank=0 of size=4, on locale.id=0
Hello, Chapel! This is MPI rank=1 of size=4, on locale.id=0
Hello, Chapel! This is MPI rank=2 of size=4, on locale.id=0
Hello, Chapel! This is MPI rank=3 of size=4, on locale.id=0
`

Unfortunately, I'm still missing a piece -- while this works here, it doesn't work under start_test, where mpirun fails to start. Not quite sure what is going on...

**Comment by npadmana** [comment]
@ben-albrecht -- added in the command line argument. start_test still doesn't work with this --- I think it might have to do with something about the environment in which the test runs.  But I'm now out of my depth....

File: test/modules/packages/mpi/multilocale/hello-chapel-nl-1/SKIPIF (hunk 1 above)
Comment (ben-albrecht) [inline test/modules/packages/mpi/multilocale/hello-chapel-nl-1/SKIPIF]: Now that #4216 is merged, you should be able to check the value of `CHPL_COMM_SUBSTRATE` directly after a rebase.

**Comment by npadmana** [comment]
@ben-albrecht -- I just updated the SKIPIF, and verified that it ran with start_test in my installation.

File: modules/packages/MPI.chpl [hunk 2]
```diff
@@ -201,7 +201,8 @@ module MPI {
     coforall loc in Locales do on loc {
       // TODO : Need a gasnet barrier here???
       var provided : c_int;
-      C_MPI.MPI_Init_thread(0,0,MPI_THREAD_MULTIPLE,provided);
+      var ptrnull : c_void_ptr;
+      C_MPI.MPI_Init_thread(ptrnull, ptrnull, MPI_THREAD_MULTIPLE,provided);
```
Comment (ben-albrecht) [inline modules/packages/MPI.chpl]: this works?

**Comment by ben-albrecht** [comment]
Closing this until we get #4214 merged.

**Pull request merged**